### ISSUE TITLE
DNS tracer is increasing CPU load on CoreDNS pods

### ISSUE THREAD

**Issue by matthyx** [inspektor-gadget/inspektor-gadget#3705]
## Description

DNS tracer doesn't filter on container namespace and therefore parses all DNS packets happening on a Kubernetes node.

## Impact

Our agent and colocated CoreDNS pods see an increased CPU consumption when using DNS tracers. Excluding CoreDNS containers doesn't change the situation.

## Environment and steps to reproduce

1. **Set-up**:
  a. start a kubernetes cluster (kind is fine)
  b. helm repo add coredns https://coredns.github.io/helm
  c. clone https://github.com/northflank-guides/dns-performance-tests.git
  d. add label `coredns-test: coredns` to a node
  e. helm --namespace=coredns-test install coredns coredns/coredns --create-namespace -f dns-performance-tests/coredns-values.yaml
  f. cd dns-performance-tests/kubernetes-client-script/
  g. npm install
  h. node index.js
2. **Task**:
  a. start dnsperf following https://github.com/northflank-guides/dns-performance-tests/tree/main?tab=readme-ov-file#dnsperf
  b. fix image in deployment using `guessi/dnsperf`
  c. reduce dnsperf `MAX_QPS` to make sure container isn't CPU throttled
4. **Action(s)**:
  a. check coredns CPU
  b. start DNS tracer and limit to a NS `kubectl gadget trace dns -n default`
5. **Error**:
  a. check core CPU increase by 10-20%

## Expected behavior

Excluding CoreDNS containers should be working and reducing CPU load.

## Additional information

Slack discussion: https://kubernetes.slack.com/archives/CSYL75LF6/p1720705023758379
WIP PR: https://github.com/inspektor-gadget/inspektor-gadget/pull/3159

**Comment by mqasimsarfraz** [comment]
@matthyx I am trying to reproduce the issue and wanted to clarify if you are running any workloads in the `default` namespace? The reason I am asking is, if I have a pod with `HostNetwork=true` in the namespace we want to trace I see much more spike in CoreDNS cpu (`700m` -> `950m`) compared to (`700m` -> `760m`) if we don't have anything installed.

**Comment by matthyx** [comment]
I didn't have anything installed for the reproducer, but in real clusters there's and indeed the spikes are higher

**Comment by matthyx** [comment]
@mauriciovasquezbernal did this slide out of the v0.36.0 milestone?

**Comment by mqasimsarfraz** [comment]
@matthyx No, I will hopefully propose an initial fix soon! I am targeting v0.37.0 milestone!

**Comment by mqasimsarfraz** [comment]
I wanted to share an update regarding this issue. The `10%` CPU overhead seem to be coming from `sockerenricher`. Under heaving DNS traffic (using https://github.com/northflank-guides/dns-performance-tests.git) `socketenricher` seem to [inserting sockets](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/socketenricher/bpf/socket-enricher.bpf.c#L44) for each DNS request regardless of the network namespace. It seems the overhead of inserting sockets increased because of https://github.com/inspektor-gadget/inspektor-gadget/pull/2684 as confirmed below:

### v0.33.0 (with 2684)
```
PROGID     TYPE                                     NAME                                     PID                                     COMM                                                 RUNTIME RUNCOUNT                        MAPMEMORY MAPCOUNT  
1004       Kprobe                                   ig_udp_sendmsg                           226586                                  gadgettracerman                                 136.667556ms 49993                            130.9MiB 4         
1003       Kprobe                                   ig_udp6_sendmsg                          226586                                  gadgettracerman                                  81.068157ms 24997                            130.9MiB 4         
72         Kprobe                                   __x64_sys_socke                          1810                                    wdavdaemon                                          348.33µs 65                               6.157MiB 7         
955        Tracing                                  ig_top_ebpf_it                           226510                                  ig                                                 290.693µs 7954                                 8KiB 1         
80         Kprobe                                   __x64_sys_accep                          1810                                    wdavdaemon                                         245.299µs 75                               6.157MiB 7  
```

### v0.32.0 (w/o 2684)
```
PROGID     TYPE                                     NAME                                     PID                                     COMM                                                 RUNTIME RUNCOUNT                        MAPMEMORY MAPCOUNT  
947        Kprobe                                   ig_udp_sendmsg                           223537                                  gadgettracerman                                  29.941133ms 50000                            2.252MiB 1         
946        Kprobe                                   ig_udp6_sendmsg                          223537                                  gadgettracerman                                  25.963593ms 25000                            2.252MiB 1         
72         Kprobe                                   __x64_sys_socke                          1810                                    wdavdaemon                                         320.373µs 56                               6.157MiB 7         
659        Tracing                                  ig_top_ebpf_it                           131895                                  ig                                                 319.342µs 8032                                 8KiB 1         
942        Kprobe                                   ig_free_ipv4_e                           223537                                  gadgettracerman                                    242.703µs 58                               2.252MiB 2  
```
`RUNTIME` 29.94ms (v0.32.0) vs 136.66ms (v0.33.0). For now, we need 2 independent optimizations 1) Skip inserting the socket if it already exists (I will open the PR soon) 2) optimize `paths`

Also, we need 3rd optimization regarding skip copies of packets on host network namespace otherwise having a pod with `hostNetwork=true` in a Kubernetes namespace make filtering of coredns pods redundant!

**Comment by alban** [comment]
> 2) optimize paths

I assume the slow part is to resolve the path. i.e. dereferencing some pointers for each component of the path. I assume copying the string is fast.

Falco has a similar feature to get the `cwd` but instead of resolving the path on demand, it keeps a map containing the path for each process. Then, it hooks on `chdir` (and `clone` IIRC) to update the path when it can change. In this way, the path would not need to be resolved for each network packet, but only for each `chdir` and `clone`, which should more a lot more rare.

That can give slightly different results if another process move an underlying path component between a chdir and a dns request, but I think we can live with that.

We can either use a hash map and cleanup the entries on `tracepoint/sched/sched_process_exit`, or use a task storage map (`BPF_MAP_TYPE_TASK_STORAGE`, Linux >= 5.11, [4cf1bc1f1045](https://github.com/torvalds/linux/commit/4cf1bc1f10452065a29d576fc5693fc4fab5b919)) so that the entries would be automatically cleaned up.

Similarly, we can store the `exepath` in a hash map or a task storage map, and update it when it changes (at execve, or [prctl(PR_SET_MM, PR_SET_MM_EXE_FILE, ...)](https://www.man7.org/linux/man-pages/man2/PR_SET_MM_EXE_FILE.2const.html))

**Comment by mqasimsarfraz** [comment]
https://github.com/inspektor-gadget/inspektor-gadget/pull/3923 should help with some performance improvement but as next step I would like to look into:
> we need 3rd optimization regarding skip copies of packets on host network namespace otherwise having a pod with hostNetwork=true in a Kubernetes namespace make filtering of coredns pods redundant!

Mainly I want to look into a generic way to exclude pods using host network!
